mac80211: realtek: rtw88: add support for RTL8723DS and RTL8723DU

RadxaNaoki opened the pull request:
add Realtek RTL8723DS and RTL8723DU support to rtw88 package.

currently it doesn't work on target devices. (8723DS for ROCK Pi S, 8723DU for ROCK Pi E)
kernel modules are loaded, but nothing happen. wlan device is not populated.

testers wanted.

RadxaNaoki commented:
similar error

https://forum.openwrt.org/t/orange-pi-prime-and-soc-wireless-rtl8723bs/195221
https://forum.openwrt.org/t/openwrt-support-for-rbr40-rbs40-got-rbr50-working/203084

hauke commented:
Your probably need this patch for the Radxa ROCK Pi S:
https://github.com/torvalds/linux/commit/12c3ec878cbe3709782e85b88124abecc3bb8617

Do you get an error message?

MilkVNaoki commented:
as you mentioned https://github.com/openwrt/openwrt/pull/15933#discussion_r1685766185, it's included.

no error message. just wlan0 is missing. phy0 is populated.

I tried linux-next on ROCK Pi S/E, wlan0 was populated as usual.

RadxaNaoki commented:
as you mentioned https://github.com/openwrt/openwrt/pull/15933#discussion_r1685766185, it's included.

no error message. just wlan0 is missing. phy0 is populated.

I tried linux-next on ROCK Pi S/E, wlan0 was populated as usual.

hauke commented:
mac80211 was updated recently to version 6.9.9, see 1bfcc1ea8a78ae834055bf554aa45720c09ea919
Could you retry with that code please.

RadxaNaoki commented:
I already tried mac80211 from 6.9.9. it's same.

RadxaNaoki commented:
I (tempolary) pushed my testing branch. please ignore 1st and 2nd commit.

kernel modules are loaded, but wlan0 is not populated.

on Pi S,
```
root@OpenWrt:/# dmesg | grep 8723
[   15.002624] rtw_8723ds mmc2:0001:1: Firmware version 48.0.0, H2C version 0
[   16.763813] rtw_8723ds mmc2:0001:1: efuse MAC invalid, using random
root@OpenWrt:/# lsmod
cdc_ether              12288  2 rndis_host,cdc_ncm
cdc_ncm                20480  0 
cfg80211              327680  2 rtw88_core,mac80211
cmac                   12288  0 
compat                 12288  3 rtw88_8723ds,mac80211,cfg80211
crc_ccitt              12288  1 ppp_async
drbg                   24576  0 
geniv                  12288  1 seqiv
gpio_button_hotplug    16384  0 
hmac                   12288  0 
jitterentropy_rng      16384  0 
mac80211              630784  2 rtw88_sdio,rtw88_core
nf_conntrack           86016  7 nft_redir,nft_nat,nft_masq,nft_flow_offload,nft_ct,nf_nat,nf_flow_table
nf_defrag_ipv4         12288  1 nf_conntrack
nf_defrag_ipv6         16384  1 nf_conntrack
nf_flow_table          28672  2 nf_flow_table_inet,nft_flow_offload
nf_flow_table_inet     12288  0 
nf_log_syslog          16384  0 
nf_nat                 32768  4 nft_redir,nft_nat,nft_masq,nft_chain_nat
nf_reject_ipv4         12288  2 nft_reject_ipv4,nft_reject_inet
nf_reject_ipv6         12288  2 nft_reject_ipv6,nft_reject_inet
nf_tables             192512156 nft_fib_inet,nf_flow_table_inet,nft_reject_ipv6,nft_reject_ipv4,nft_reject_inet,nft_reject,nft_redir,nft_quota,nft_numgen,nft_nat,nft_masq,nft_log,nft_limit,nft_hash,nft_flow_offload,nft_fib_ipv6,nft_fib_ipv4,nft_fib,nft_ct,nft_chain_nat
nfnetlink              16384  1 nf_tables
nft_chain_nat          12288  2 
nft_ct                 16384  3 
nft_fib                12288  3 nft_fib_inet,nft_fib_ipv6,nft_fib_ipv4
nft_fib_inet           12288  0 
nft_fib_ipv4           12288  1 nft_fib_inet
nft_fib_ipv6           12288  1 nft_fib_inet
nft_flow_offload       12288  0 
nft_hash               12288  0 
nft_limit              12288  5 
nft_log                12288  0 
nft_masq               12288  1 
nft_nat                12288  0 
nft_numgen             12288  0 
nft_quota              12288  0 
nft_redir              12288  0 
nft_reject             12288  3 nft_reject_ipv6,nft_reject_ipv4,nft_reject_inet
nft_reject_inet        12288  2 
nft_reject_ipv4        12288  0 
nft_reject_ipv6        12288  0 
ppp_async              16384  0 
ppp_generic            40960  3 pppoe,ppp_async,pppox
pppoe                  20480  0 
pppox                  12288  1 pppoe
rndis_host             12288  0 
rng                    12288  3 jitterentropy_rng,drbg,geniv
rtw88_8723d            40960  1 rtw88_8723ds
rtw88_8723ds           12288  0 
rtw88_core            135168  2 rtw88_8723d,rtw88_sdio
rtw88_sdio             20480  1 rtw88_8723ds
seqiv                  12288  0 
sha3_generic           12288  0 
sha512_arm64           16384  0 
sha512_generic         12288  0 
slhc                   12288  1 ppp_generic
usbnet                 24576  3 rndis_host,cdc_ncm,cdc_ether
root@OpenWrt:/# ls /sys/class/ieee80211/
phy0
root@OpenWrt:/# ls /sys/class/net
br-lan  eth0    lo
```

on Pi E,
```
root@OpenWrt:/# dmesg | grep 8723
[    5.668840] rtw_8723du 2-1:1.2: Firmware version 48.0.0, H2C version 0
[    6.817727] usbcore: registered new interface driver rtw_8723du
root@OpenWrt:/# lsmod
cdc_ether              12288  2 rndis_host,cdc_ncm
cdc_ncm                20480  0 
cfg80211              327680  2 rtw88_core,mac80211
cmac                   12288  0 
compat                 12288  3 rtw88_8723du,mac80211,cfg80211
crc_ccitt              12288  1 ppp_async
drbg                   24576  0 
geniv                  12288  1 seqiv
gpio_button_hotplug    16384  0 
hmac                   12288  0 
jitterentropy_rng      16384  0 
mac80211              630784  2 rtw88_usb,rtw88_core
nf_conntrack           86016  7 nft_redir,nft_nat,nft_masq,nft_flow_offload,nft_ct,nf_nat,nf_flow_table
nf_defrag_ipv4         12288  1 nf_conntrack
nf_defrag_ipv6         16384  1 nf_conntrack
nf_flow_table          28672  2 nf_flow_table_inet,nft_flow_offload
nf_flow_table_inet     12288  0 
nf_log_syslog          16384  0 
nf_nat                 32768  4 nft_redir,nft_nat,nft_masq,nft_chain_nat
nf_reject_ipv4         12288  2 nft_reject_ipv4,nft_reject_inet
nf_reject_ipv6         12288  2 nft_reject_ipv6,nft_reject_inet
nf_tables             192512208 nft_fib_inet,nf_flow_table_inet,nft_reject_ipv6,nft_reject_ipv4,nft_reject_inet,nft_reject,nft_redir,nft_quota,nft_numgen,nft_nat,nft_masq,nft_log,nft_limit,nft_hash,nft_flow_offload,nft_fib_ipv6,nft_fib_ipv4,nft_fib,nft_ct,nft_chain_nat
nfnetlink              16384  1 nf_tables
nft_chain_nat          12288  2 
nft_ct                 16384  4 
nft_fib                12288  3 nft_fib_inet,nft_fib_ipv6,nft_fib_ipv4
nft_fib_inet           12288  0 
nft_fib_ipv4           12288  1 nft_fib_inet
nft_fib_ipv6           12288  1 nft_fib_inet
nft_flow_offload       12288  0 
nft_hash               12288  0 
nft_limit              12288  5 
nft_log                12288  0 
nft_masq               12288  1 
nft_nat                12288  0 
nft_numgen             12288  0 
nft_quota              12288  0 
nft_redir              12288  0 
nft_reject             12288  3 nft_reject_ipv6,nft_reject_ipv4,nft_reject_inet
nft_reject_inet        12288  2 
nft_reject_ipv4        12288  0 
nft_reject_ipv6        12288  0 
ppp_async              16384  0 
ppp_generic            40960  3 pppoe,ppp_async,pppox
pppoe                  20480  0 
pppox                  12288  1 pppoe
rndis_host             12288  0 
rng                    12288  3 jitterentropy_rng,drbg,geniv
rtw88_8723d            40960  1 rtw88_8723du
rtw88_8723du           12288  0 
rtw88_core            135168  2 rtw88_8723d,rtw88_usb
rtw88_usb              16384  1 rtw88_8723du
seqiv                  12288  0 
sha3_generic           12288  0 
sha512_arm64           16384  0 
sha512_generic         12288  0 
slhc                   12288  1 ppp_generic
usbnet                 24576  3 rndis_host,cdc_ncm,cdc_ether
root@OpenWrt:/# ls /sys/class/ieee80211/
phy0
root@OpenWrt:/# ls /sys/class/net/
br-lan  eth0    eth1    lo
```

RadxaNaoki commented:
ah, I forgot to add some patches for Pi S. I'll re-test with them. (but result should be same, I already tried)

hauke commented:
If the `phy0` device is created the device was probed. Maybe logread has some information. 
Did you run `wifi detect`?

RadxaNaoki commented:
`wifi detect` seems to be obsoleted by `wifi config`. it doesn't output anything, I run `wifi status`.

here is output from Pi S:
[pi-s-logread.txt](https://github.com/user-attachments/files/16355097/pi-s-logread.txt)
[pi-s-wifi-status.txt](https://github.com/user-attachments/files/16355098/pi-s-wifi-status.txt)

RadxaNaoki commented:
`iw phy0 info` works.
there is /etc/config/wireless.
I removed disabled option, but I cannot see `OpenWrt` ssid AP.

RadxaNaoki commented:
oh... some errors about netifd: radio9

[pi-e-logread.txt](https://github.com/user-attachments/files/16355267/pi-e-logread.txt)

RadxaNaoki commented:
I installed hostapd and related packages, then AP mode is working! I can see phy0-ap0 device by `ip` command.

I still cannot see `wlan0`, is it normal?

RadxaNaoki commented:
I cannot believe it but https://github.com/openwrt/openwrt/pull/15910/commits/94d075934151fe3ef689bf17eb2ae02431c5c2f2 makes things work

RadxaNaoki commented:
> I cannot believe it but [94d0759](https://github.com/openwrt/openwrt/commit/94d075934151fe3ef689bf17eb2ae02431c5c2f2) makes things work
> 
> maybe I should enable hostapd in .config...?

ok, I understood how to add package in image. wifi was working since beggining (with ucode rtnl fix), I need to understand I have to use `CONFIG_PACKAGE_foo=y` in .config.

it should be ok now

hauke commented on package/firmware/linux-firmware/realtek.mk:
```diff
@@ -107,6 +107,9 @@ Package/rtl8723de-firmware = $(call Package/firmware-default,RealTek RTL8723DE f
 	$(INSTALL_DATA) $(PKG_BUILD_DIR)/rtw88/rtw8723d_fw.bin $(1)/lib/firmware/rtw88
+	$(INSTALL_DIR) $(1)/lib/firmware/rtl_bt
+	$(INSTALL_DATA) $(PKG_BUILD_DIR)/rtl_bt/rtl8723d_config.bin $(1)/lib/firmware/rtl_bt
+	$(INSTALL_DATA) $(PKG_BUILD_DIR)/rtl_bt/rtl8723d_fw.bin $(1)/lib/firmware/rtl_bt
```
Please add an extra package for these two files.

RadxaNaoki replied:
I see. I'll do it separately.

hauke commented:
Please install `wpad-basic-mbedtls`. You can add it as a default package for the device like this:
```
diff --git a/target/linux/rockchip/image/armv8.mk b/target/linux/rockchip/image/armv8.mk
index 063dc6fd64..f2448f2cf6 100644
--- a/target/linux/rockchip/image/armv8.mk
+++ b/target/linux/rockchip/image/armv8.mk
@@ -58,6 +58,7 @@ define Device/radxa_rock-pi-4a
   SUPPORTED_DEVICES := radxa,rockpi4a radxa,rockpi4
   UBOOT_DEVICE_NAME := rock-pi-4-rk3399
   IMAGE/sysupgrade.img.gz := boot-common | boot-script | pine64-img | gzip | append-metadata
+  DEVICE_PACKAGES := wpad-basic-mbedtls kmod-rtw88-8723ds
 endef
 TARGET_DEVICES += radxa_rock-pi-4a
 
```

RadxaNaoki committed:
rockchip: add RTL8723DS support for Radxa ROCK Pi S

Radxa ROCK Pi S has a RTL8723DS Wi-Fi 4 on-board device. enable it.

Signed-off-by: FUKAUMI Naoki <naoki@radxa.com>
Link: https://github.com/openwrt/openwrt/pull/15910
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

RadxaNaoki committed:
rockchip: add RTL8723DU support for Radxa ROCK Pi E v3.0

Radxa ROCK Pi E v3.0 has a RTL8723DU Wi-Fi 4 on-board device. enable
it.

Signed-off-by: FUKAUMI Naoki <naoki@radxa.com>
Link: https://github.com/openwrt/openwrt/pull/15910
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

RadxaNaoki commented:
> Please install `wpad-basic-mbedtls`. You can add it as a default package for the device like this:
> 
> ```
> diff --git a/target/linux/rockchip/image/armv8.mk b/target/linux/rockchip/image/armv8.mk
> index 063dc6fd64..f2448f2cf6 100644
> --- a/target/linux/rockchip/image/armv8.mk
> +++ b/target/linux/rockchip/image/armv8.mk
> @@ -58,6 +58,7 @@ define Device/radxa_rock-pi-4a
>    SUPPORTED_DEVICES := radxa,rockpi4a radxa,rockpi4
>    UBOOT_DEVICE_NAME := rock-pi-4-rk3399
>    IMAGE/sysupgrade.img.gz := boot-common | boot-script | pine64-img | gzip | append-metadata
> +  DEVICE_PACKAGES := kmod-rtw88-8723ds wpad-basic-mbedtls
>  endef
>  TARGET_DEVICES += radxa_rock-pi-4a
>  
> ```

done in last 2 commit (I separated "add" and "use" changes)

hauke commented on target/linux/rockchip/image/armv8.mk:
```diff
@@ -157,7 +157,7 @@ define Device/radxa_rock-pi-s
   BOOT_SCRIPT := rock-pi-s
   UBOOT_DEVICE_NAME := rock-pi-s-rk3308
-  DEVICE_PACKAGES := kmod-usb-net-cdc-ncm kmod-usb-net-rndis
+  DEVICE_PACKAGES := kmod-rtw88-8723ds kmod-usb-net-cdc-ncm kmod-usb-net-rndis wpad-basic-mbedtls
```
Please move this to the last patch

RadxaNaoki commented:
rebased.

first two changes should be merged by https://github.com/openwrt/openwrt/pull/15984 and https://github.com/openwrt/openwrt/pull/16029

RadxaNaoki commented:
rebased

hauke commented:
A description is missing in the commit message of the last 2 patches:
```
Run source .github/workflows/scripts/ci_helpers.sh
=== Checking commit '9400b054b4ba78afdc6c91765533ebb6cf4c629f'
Author name (FUKAUMI Naoki) seems ok
Commit subject line seems ok (rockchip: add RTL8723DU support for Radxa ROCK Pi E)
Signed-off-by match author
Missing commit message. Please describe your changes
=== Checking commit '0a978243a76acddc63c1889632b197c7c0c4f518'
Author name (FUKAUMI Naoki) seems ok
Commit subject line seems ok (rockchip: add RTL8723DS support for Radxa ROCK Pi S)
Signed-off-by match author
Missing commit message. Please describe your changes
=== Checking commit '07151ca1dcf7d9e79e9ce178cbe69f82a280b492'
Author name (FUKAUMI Naoki) seems ok
Commit subject line seems ok (mac80211: realtek: rtw88: add support for RTL8723DS and RTL8723DU)
Signed-off-by match author
add Realtek RTL8723DS and RTL8723DU support to rtw88 package.

A commit message exists
Error: Process completed with exit code 1.
```

RadxaNaoki commented:
> A description is missing in the commit message of the last 2 patches:

thank you, fixed!

hauke commented on package/kernel/mac80211/realtek.mk:
```diff
@@ -201,6 +204,15 @@ define KernelPackage/rtw88-pci
+define KernelPackage/rtw88-sdio
+  $(call KernelPackage/mac80211/Default)
+  TITLE:=Realtek RTW88 SDIO chips support
+  DEPENDS+= +kmod-rtw88
```
```suggestion
  DEPENDS+= +kmod-mmc +kmod-rtw88
```

The CI system found the following error:
```
Package kmod-rtw88-sdio is missing dependencies for the following libraries:
mmc_core.ko
make[3]: *** [Makefile:405: /__w/openwrt/openwrt/openwrt/bin/targets/malta/be/packages/kmod-rtw88-sdio_6.6.43.6.9.9-r1_mips_24kc.ipk] Error 1
time: package/kernel/mac80211/regular/compile#798.38#118.47#425.70
Error: Process completed with exit code 2.
```

RadxaNaoki replied:
fixed, thanks

RadxaNaoki replied:
strange, mmc_core and mmc_block kernel modules are not loaded (not exist in rootfs)

RadxaNaoki replied:
`grep -r sdio target/linux/malta` doesn't return anything. is CI system use special configuration?

hauke replied:
For the rockchip target the MMC core is build into the kernel, then the kmod-mmc is empty. For the malta target the MMC core is not build into the kernel, then it is build as a module and packaged in kmod-mmc.

RadxaNaoki replied:
I see. then, is my last fix okay?

RadxaNaoki committed:
mac80211: realtek: rtw88: add support for RTL8723DS and RTL8723DU

add Realtek RTL8723DS and RTL8723DU support to rtw88 package.

Signed-off-by: FUKAUMI Naoki <naoki@radxa.com>
Link: https://github.com/openwrt/openwrt/pull/15910
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>
